### PR TITLE
Add Box/Unbox as first class NIR Ops

### PR DESCRIPTION
**Pull request by LukasKellenberger**

### CODE REVIEW

File: nir/src/main/scala/scala/scalanative/nir/Tags.scala [hunk 1]
```diff
@@ -154,6 +154,8 @@ object Tags {
   final val CopyOp       = 1 + IsOp
   final val SizeofOp     = 1 + CopyOp
   final val ClosureOp    = 1 + SizeofOp
+  final val Box          = 1 + ClosureOp
+  final val Unbox        = 1 + Box
```
Comment (densh) [inline nir/src/main/scala/scala/scalanative/nir/Tags.scala]: For naming consistency: `BoxOp`, `UnboxOp`

File: nir/src/main/scala/scala/scalanative/nir/Ops.scala [hunk 2]
```diff
@@ -46,6 +49,9 @@ object Op {
   final case class Conv(conv: nir.Conv, ty: Type, value: Val)     extends Pure
   final case class Select(cond: Val, thenv: Val, elsev: Val)      extends Pure
 
+  final case class Box(code: Char, obj: Val)   extends Op
+  final case class Unbox(code: Char, obj: Val) extends Op
```
Comment (densh) [inline nir/src/main/scala/scala/scalanative/nir/Ops.scala]: 1. Those ops belong to the high-level group of ops below.
2. They should take a box type, and not a character code.

File: nir/src/main/scala/scala/scalanative/nir/Shows.scala [hunk 3]
```diff
@@ -120,6 +120,11 @@ object Shows {
     case Op.Select(cond, thenv, elsev) =>
       sh"select $cond, $thenv, $elsev"
 
+    case Op.Box(code, obj) =>
+      sh"box $code $obj"
+    case Op.Unbox(code, obj) =>
+      sh"unbox $code $obj"
```
Comment (densh) [inline nir/src/main/scala/scala/scalanative/nir/Shows.scala]: 1. Move them below to the high-level group.
2. `box[$ty] $value` and `unbox[$ty] $obj`

File: tools/src/main/scala/scala/scalanative/optimizer/pass/GlobalValueNumbering.scala [hunk 4]
```diff
@@ -286,6 +292,9 @@ object GlobalValueNumbering extends PassCompanion {
         case Conv(conv, ty, value)      => Seq("Conv", ty, value)
         case Select(cond, thenv, elsev) => Seq("Select", cond, thenv, elsev)
 
+        case Box(code, obj)   => Seq("Box", code.toString, obj)
+        case Unbox(code, obj) => Seq("Unbox", code.toString, obj)
+
         case Field(obj, name)           => Seq("Field", obj, name)
         case Method(obj, name)          => Seq("Method", obj, name)
         case As(ty, obj)                => Seq("As", ty, obj)
```
Comment (densh) [inline tools/src/main/scala/scala/scalanative/optimizer/pass/GlobalValueNumbering.scala]: /cc @hkapp

File: nir/src/main/scala/scala/scalanative/nir/Ops.scala (hunk 2 above)
Comment (densh) [inline nir/src/main/scala/scala/scalanative/nir/Ops.scala]: Yep, the boxed type in both cases.

File: util/src/main/scala/scala/scalanative/util/Show.scala [hunk 5]
```diff
@@ -68,6 +68,7 @@ object Show {
 
   implicit def showResult[R <: Result]: Show[R] = apply(identity)
   implicit def showString[T <: String]: Show[T] = apply(Show.Str(_))
+  implicit def showChar[T <: Char]: Show[T]     = apply(c => Show.Str(c.toString))
```
Comment (densh) [inline util/src/main/scala/scala/scalanative/util/Show.scala]: Not needed any longer.

File: tools/src/main/scala/scala/scalanative/optimizer/pass/GlobalValueNumbering.scala [hunk 6]
```diff
@@ -180,6 +180,12 @@ object GlobalValueNumbering extends PassCompanion {
           case (Sizeof(tyA), Sizeof(tyB)) =>
             eqType(tyA, tyB)
 
+          case (Box(codeA, objA), Box(codeB, objB)) =>
+            codeA == codeB && eqVal(objA, objB)
+
+          case (Unbox(codeA, objA), Unbox(codeB, objB)) =>
+            codeB == codeB && eqVal(objA, objB)
```
Comment (hkapp) [inline tools/src/main/scala/scala/scalanative/optimizer/pass/GlobalValueNumbering.scala]: `codeA == codeB`

**Comment by hkapp** [comment]
It seems like the only uses of `Boxes.box` are of the form `Boxes.box(genPrimCode(tpe))`. Why not have `Boxes.box` be of type `Map[Type,Type]`, reducing the uses to `Boxes.box(tpe)` ?

File: tools/src/main/scala/scala/scalanative/optimizer/pass/GlobalValueNumbering.scala (hunk 6 above)
Comment (densh) [inline tools/src/main/scala/scala/scalanative/optimizer/pass/GlobalValueNumbering.scala]: Good catch!

File: nir/src/main/scala/scala/scalanative/nir/Ops.scala [hunk 7]
```diff
@@ -56,4 +59,7 @@ object Op {
   final case class Copy(value: Val)                                extends Op
   final case class Sizeof(ty: Type)                                extends Op
   final case class Closure(ty: Type, fun: Val, captures: Seq[Val]) extends Op
-}
+  final case class Box(ty: Type, obj: Val)                         extends Op
+  final case class Unbox(ty: Type, obj: Val)                       extends Op
+
+}
```
Comment (densh) [inline nir/src/main/scala/scala/scalanative/nir/Ops.scala]: Needs newline in the end of file.

File: nir/src/main/scala/scala/scalanative/nir/Types.scala [hunk 8]
```diff
@@ -50,3 +50,21 @@ object Type {
   final case class Trait(name: Global)  extends RefKind with Named
   final case class Module(name: Global) extends RefKind with Named
 }
+
+object Boxes {
+
+  val unbox = Map[Type, Type](
```
Comment (densh) [inline nir/src/main/scala/scala/scalanative/nir/Types.scala]: This one should go straight to `object Type`.

File: nscplugin/src/main/scala/scala/scalanative/nscplugin/NirTypeEncoding.scala [hunk 9]
```diff
@@ -117,6 +117,30 @@ trait NirTypeEncoding { self: NirCodeGen =>
...
+      case UIntClass =>
+        nir.Type.Class(nir.Global.Top("scala.scalanative.native.UInt"))
+      case IntClass => nir.Type.Class(nir.Global.Top("java.lang.Integer"))
+      case ULongClass =>
+        nir.Type.Class(nir.Global.Top("scala.scalanative.native.ULong"))
+      case LongClass   => nir.Type.Class(nir.Global.Top("java.lang.Long"))
+      case FloatClass  => nir.Type.Class(nir.Global.Top("java.lang.Float"))
+      case DoubleClass => nir.Type.Class(nir.Global.Top("java.lang.Double"))
+      case _           => unsupported("Box type must be primitive type.")
+    }
```
Comment (densh) [inline nscplugin/src/main/scala/scala/scalanative/nscplugin/NirTypeEncoding.scala]: Mix between aligned and unaligned arrows looks pretty bad for this match. Lets add newline after every `=>` in match cases. Unfortunately scalafmt can't figure this out on its own yet.

File: tools/src/main/scala/scala/scalanative/optimizer/Driver.scala [hunk 10]
```diff
@@ -22,6 +22,7 @@ object Driver {
   def apply(config: tools.Config): Driver =
     new Impl(
       Seq(pass.GlobalBoxingElimination,
+          pass.BoxingLowering,
```
Comment (densh) [inline tools/src/main/scala/scala/scalanative/optimizer/Driver.scala]: This one should go right before AsLowering.

File: tools/src/main/scala/scala/scalanative/optimizer/Pass.scala [hunk 11]
```diff
@@ -135,6 +135,11 @@ trait Pass {
     case Op.Select(v1, v2, v3) =>
       Op.Select(txVal(v1), txVal(v2), txVal(v3))
 
+    case Op.Box(code, obj) =>
+      Op.Box(code, txVal(obj))
+    case Op.Unbox(code, obj) =>
+      Op.Unbox(code, txVal(obj))
+
```
Comment (densh) [inline tools/src/main/scala/scala/scalanative/optimizer/Pass.scala]: Similarly to case class definitions for ops we group cases into low-level and high-level. Those two should go in the end of this match.

File: tools/src/main/scala/scala/scalanative/optimizer/pass/GlobalValueNumbering.scala [hunk 12]
```diff
@@ -286,6 +292,9 @@ object GlobalValueNumbering extends PassCompanion {
         case Conv(conv, ty, value)      => Seq("Conv", ty, value)
         case Select(cond, thenv, elsev) => Seq("Select", cond, thenv, elsev)
 
+        case Box(code, obj)   => Seq("Box", code.toString, obj)
+        case Unbox(code, obj) => Seq("Unbox", code.toString, obj)
```
Comment (densh) [inline tools/src/main/scala/scala/scalanative/optimizer/pass/GlobalValueNumbering.scala]: Should go the end similarly to the one above.

File: nscplugin/src/main/scala/scala/scalanative/nscplugin/NirTypeEncoding.scala (hunk 9 above)
Comment (LukasKellenberger) [inline nscplugin/src/main/scala/scala/scalanative/nscplugin/NirTypeEncoding.scala]: Should I add a newline even if `scalafmt --test` fails ?

**Comment by densh** [comment]
LGTM, thanks @LukasKellenberger !

**Pull request merged**